examples/greeting/public/plugin.ts

```diff
@@ -0,0 +1,83 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Plugin } from '../../../src/core/public';
+
+export interface GreetingDefinition {
+  id: string;
+  salutation: string;
+  punctuation: string;
+}
+
+export interface Greeter {
+  greetMe: (name: string) => void;
+  label: string;
+}
+
+type GreetingProvider = (def: GreetingDefinition) => Greeter;
+
+const defaultGreetingProvider: GreetingProvider = (def: GreetingDefinition) => ({
+  greetMe: (name: string) => alert(`${def.salutation} ${name}${def.punctuation}`),
+  label: def.id,
+});
+
+export interface GreetingStart {
+  /**
+   * This function should be used if the value of `id` is not known at compile time. Usually
+   * this is because `id` has been persisted somewhere. If the value of `id` is known at
+   * compile time (e.g. `greeting.getGreeter(CASUAL_GREETER))`) developers should prefer accessing
+   * off the plugin that registered `CasualGreeter`, like `customGreetings.getCasualGreeter()`, because:
+   *  - makes it more explicit you should add `customGreetings` to your plugin dependency list.
+   *  - don't need to handle the possibility of `undefined` if it's a required dependency.
+   *  - can get more specialized types in certain situations (e.g. `interface CustomGreeter` vs `Greeter` -
+   * does not apply to this example but many real world examples are like this.)
+   */
+  getGreeter: (id: string) => Greeter;
+  /**
+   * Returns an array of all registered greeters.
+   */
+  getGreeters: () => Greeter[];
+}
+
+export interface GreetingSetup {
+  setCustomProvider: (customProvider: GreetingProvider) => void;
+  registerGreetingDefinition: (greetingDefinition: GreetingDefinition) => () => Greeter;
+}
+
+export class GreetingPlugin implements Plugin<GreetingSetup, GreetingStart> {
+  private greetingDefinitions: { [key: string]: GreetingDefinition } = {};
+  private greetingProvider: GreetingProvider = defaultGreetingProvider;
+
+  setup = () => ({
+    setCustomProvider: (customProvider: GreetingProvider) =>
+      (this.greetingProvider = customProvider),
+    registerGreetingDefinition: (greetingDefinition: GreetingDefinition) => {
+      this.greetingDefinitions[greetingDefinition.id] = greetingDefinition;
+      return () => this.greetingProvider(greetingDefinition);
```

Comment:
Doesn't returning `() => this.greetingProvider(greetingDefinition)` from within the `setup` lifecycle potentially cause issues if this is called before another plugin has the chance to call `setCustomProvider`?

Reply:
yea, good catch. 

how about this change? (sorry I shouldn't have squashed the second commit, force of habit).  

it ensures that an error is thrown if the accessor is used before this plugin's start lifecycle.